Signal/src/Profiles/OWSProfileManager.m

```diff
@@ -671,156 +704,181 @@ - (void)setContactRecipientIds:(NSArray<NSString *> *)contactRecipientIds
-                        || ![currentUserProfile.profileKey isEqual:userProfile.profileKey]) {
-                        DDLogWarn(@"%@ Ignoring avatar download for obsolete user profile.", self.tag);
-                    } else if (error) {
-                        DDLogError(@"%@ avatar download failed: %@", self.tag, error);
-                    } else if (!encryptedData) {
-                        DDLogError(@"%@ avatar encrypted data could not be read.", self.tag);
-                    } else if (!decryptedData) {
-                        DDLogError(@"%@ avatar data could not be decrypted.", self.tag);
-                    } else if (!image) {
-                        DDLogError(@"%@ avatar image could not be loaded: %@", self.tag, error);
-                    } else {
-                        [self.otherUsersProfileAvatarImageCache setObject:image forKey:userProfile.recipientId];
-
-                        userProfile.avatarFileName = fileName;
-
-                        [self saveUserProfile:userProfile];
-                    }
-                });
-            });
-        }];
-    [downloadTask resume];
+                        @synchronized(self)
+                        {
+                            [self.currentAvatarDownloads removeObject:userProfile.recipientId];
+
+                            UserProfile *currentUserProfile =
+                                [self getOrBuildUserProfileForRecipientId:userProfile.recipientId];
+                            if (currentUserProfile.profileKey.keyData.length < 1
+                                || ![currentUserProfile.profileKey isEqual:userProfile.profileKey]) {
+                                DDLogWarn(@"%@ Ignoring avatar download for obsolete user profile.", self.tag);
+                            } else if (![avatarUrlPathAtStart isEqualToString:currentUserProfile.avatarUrlPath]) {
+                                DDLogInfo(@"%@ avatar url has changed during download", self.tag);
+                                if (currentUserProfile.avatarUrlPath.length > 0) {
+                                    [self downloadAvatarForUserProfile:currentUserProfile];
+                                }
+                            } else if (error) {
+                                DDLogError(@"%@ avatar download failed: %@", self.tag, error);
+                            } else if (!encryptedData) {
+                                DDLogError(@"%@ avatar encrypted data could not be read.", self.tag);
+                            } else if (!decryptedData) {
+                                DDLogError(@"%@ avatar data could not be decrypted.", self.tag);
+                            } else if (!image) {
+                                DDLogError(@"%@ avatar image could not be loaded: %@", self.tag, error);
+                            } else {
+                                // TODO: Verify that the avatar URL hasn't changed since the download began.
```

Comment:
re: TODO, it seems like you've addressed this with your additions a few lines up?

```
 } else if (![avatarUrlPathAtStart isEqualToString:currentUserProfile.avatarUrlPath]) {
...
```

Also to nit: `currentUserProfile` isn't the best name. For a second I thought it referred to the "local" user. Maybe `latestUserProfile`? Names are hard. Take it or leave it.